### PR TITLE
Features / Groups

### LINKED ISSUE #103: Add Groups
This issue depends on #95, as groups should have invitations.

Events will be handled by another issue.

## Backend

A new table `groups` should be implemented. It will have columns: `primary key`, `title`, `description`  and `creator_id` (foreign key to `users`) and `timestamp`. 

A new table `group_members` should be implemented. It is very similar to `followers` (#94). Columns: `primary key`, `group_id` (foreign key to `groups`), `user_id` (primary key to `users`) and `timestamp`. 

Groups can have posts in it, so `posts` table should be updated with a new optional column `group_id` (foreign key to table `groups`). 

For posts in groups we'll save to the database both author `user_id` and `group_id`). For non-group posts we'll save `null` to `group_id` column. A new `groupId` field will be added to `/posts/create`. 

Several new endpoints should be implemented. It's similar to ones we have for users:

- `GET /api/groups/{id}` - get data about a specific group. It should have the following structure:

```ts
{
    id: 1,
    title: "Amazing group",
    description: "A great place for great posts",

    // similar to `follow_status`, can be `0`-not member, `1`-member and `2`-requested membership
    member_status: 0, 

    members_count: 122, 
}
```

- `GET /api/groups/{id}/posts` - get array of ids of posts of the specific group (sorted by date, newest first). 

- `GET /api/groups` - returns array of ids of all groups, sorted by member count.

- `POST /api/groups/{id}/join` - request joining the group, no request body, available to anyone who is singed up. 

- `POST /api/groups/{id}/invite` - invite user by id, should be available to logged in users who are members of the group. 

```ts
{
    user_id: 1, // id of the user to invite
}
```

- `POST /api/groups/{id}/leave` - leave the group, empty body

Chats and events will be handled in another issue

### PR DESCRIPTION
**Pull request by maximihajlov**
This pull request resolves #103. It adds groups!

Basic functionality works as expected:

Users can be invited to join group while creating it:

https://github.com/merpw/forum/assets/43277901/88d0f3c4-b938-4828-9053-191aee37b514

They can also join groups sending a request from group page:

https://github.com/merpw/forum/assets/43277901/9e27af6b-ade5-449e-a554-9c02f2fd26d9

Or they can be invited by the members of the group:

https://github.com/merpw/forum/assets/43277901/1cb538cf-85a7-408e-9e54-15acb579f891

But there's an issue that still exists. Group posts are added to `/api/posts` and available to anyone (even not members of a group). This will be fixed as a part of #99 as it requires moving posts fetching to the client side.

https://github.com/merpw/forum/assets/43277901/d8d61cbf-04d6-4a55-a141-42b48f1c33a1

### CODE REVIEW

**Comment by maximihajlov** [comment]
The frontend for most functionality is done. It was not tested with the real backend, but request and response format is the same as described in #103

Inviting the followers after creating the group is not implemented yet. According to the task description, users should be able to invite their followers to the group. The problem is that using `/api/me/followers` is not enough, because we have to filter users that are not yet in the group. 

To do so, we can make follower list public (something like `/api/groups/{id}/members`, or implement another endpoint with the necessary data, like `/api/groups/{id}/users_to_invite`.

**Comment by maximihajlov** [comment]
While making videos I found that we don't mutate group when accepting the invitation (so we have to re-open the group page to see the new fields). I've just fixed it in cf816debca2708f8052a2a64496a9766d57591bd

https://github.com/merpw/forum/assets/43277901/131884d2-d3bf-479a-9259-e1b41a975349

File: backend/forum/database/ops_groups.go [hunk 1]
```diff
@@ -61,8 +61,8 @@ func (db DB) GetGroupById(groupId int) *Group {
 	group := &Group{
 		Id: groupId,
 	}
-	err := db.QueryRow(`SELECT title, description FROM groups WHERE id = ?`,
-		groupId).Scan(&group.Title, &group.Description)
+	err := db.QueryRow(`SELECT title, description, creator_id FROM groups WHERE id = ?`,
+		groupId).Scan(&group.Title, &group.Description, &group.CreatorId)
 	if err != nil {
```
Comment (bomanviktor) [inline backend/forum/database/ops_groups.go:66]: Great idea. I forgot that you need the ID of the creator to determine if the button should be shown or not!

**Comment by bomanviktor** [comment]
Backend looking solid after you fixed the problems with the `GetTopGroups` query, and after you updated the `GetGroupMemberStatus` to take the actual `from_user_id` and `to_user_id` into consideration.

**Comment by bomanviktor** [comment]
> According to the task description, users should be able to invite their followers to the group. The problem is that using `/api/me/followers` is not enough, because we have to filter users that are not yet in the group.

Let's not have restrictions on who you can invite. It does not say that you _should not_ be able to invite someone else, just that you _should_ be able to invite one of your followers.

**Review by FinnTune — APPROVED** [review]
LGTM. Nicely done.

Not related to this but the text contrast for `Failed to upload image` is not good on light mode. 
image

Also, the message for image upload should also have "click here to upload image" added:*
image

**Review by bomanviktor — APPROVED** [review]
I looked through the frontend together with Andre, and it looks really good! Well done.

But if you can fix the contrast of the error of image upload it would be nice.

I think it should be someting like this:

`[Upload File] or drag and drop an image`

**Comment by maximihajlov** [comment]
> Not related to this but the text contrast for `Failed to upload image` is not good on light mode. 

This PR is about groups, the issue with contrast will be fixed later in #101 as well as other design imperfections. It's not perfect, there's also no loading bar. So maybe we'll make a patch PR for this.

> Also, the message for image upload should also have "click here to upload image" added

Used the same text as on GitHub. In my opinion, there's no need to make it longer. User can always hover it and see that it's clickable

Screenshot_2023-08-12-17-24-10-43_40deb401b9ffe8e1df2f1cc5ba480b12

**Comment by maximihajlov** [comment]
[Deployed](https://forum-ci.mer.pw) successfully. Thanks for the reviews!

**Pull request merged**